### PR TITLE
C++: Recognize indirect `argv` accesses as flow sources for use-use dataflow

### PR DESCRIPTION
**Pull request by jketema**
This fixes the test regression on `cpp/command-line-injection`.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll [hunk 1]
```diff
@@ -53,7 +53,7 @@ private class ArgvSource extends LocalFlowSource {
     exists(Function main, Parameter argv |
       main.hasGlobalName("main") and
       main.getParameter(1) = argv and
-      this.asExpr() = argv.getAnAccess()
+      argv.getAnAccess() in [this.asExpr(), this.asIndirectExpr()]
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll:56]: `argv.getAnAccess() = this.asIndirectExpr` is definitely the right solution. But I don't know if there are other places that expect `this.asExpr()` as well 🤔.

[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll:56]: Not in our tests, but that doesn't say a lot.

[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll:56]: 😂. Let's keep the `this.asExpr` case for now, then. Then we can do a MRVA sweep over cases like this (and the `getenv` one) in a different PR.

**Review by MathiasVP — APPROVED** [review]
LGTM!

File: cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/security/FlowSources.qll:56]: Ok, let me at least run a DCA experiment before we merge this.

**Comment by jketema** [comment]
DCA shows some new results. These seem to be due to other uses of `argv` now being identified as sources.

**Comment by MathiasVP** [comment]
> @MathiasVP I tried to express: I'm not sure if that's what we want 😄

Oh, I misunderstood your comment, then 😂. Do you mean that the new results are duplicated results due to flow going into the `argv` source?

**Comment by jketema** [comment]
Not really duplicates, more a side-effect of switching to use-use dataflow, I guess.
```cpp
	trace2_cmd_start(argv); // This argv use becomes a source
	trace2_collect_process_info();

	git_resolve_executable_dir(argv[0]); // This argv use becomes a source

	git_setup_gettext();

	initialize_the_repository();

	attr_start();

	result = cmd_main(argc, argv); // This argv use becomes a source
```
The last source is probably the correct one, as the sink in each case is reachable from `cmd_main`.

**Comment by MathiasVP** [comment]
I see. Yeah, what we'd really like to do is to mark the `argv` _parameter_ as the source of flow (which we can do now that we're using the `FlowSource` class and not the `isUserInput` predicate 🎉). But if that's too drastic of a change, would it not remove the additional results if we block flow _into_ `FlowSource`s? Then flow wouldn't go from `argv (line 1) -> argv[0] (line 4) -> argv (line 12)`, since all of those nodes wouldn't have _incoming_ flow. But we'd still get flow out of `argv (line 12)` and into the sinks we care about?

**Comment by jketema** [comment]
If we go the blocking route, there's probably a whole bunch of queries that need to be fixed up, so that's not ideal.

I'm not exactly sure how to mark the parameter. I could do `this.asParameter()`, but that doesn't give me the indirections?

**Comment by MathiasVP** [comment]
> I'm not exactly sure how to mark the parameter. I could do `this.asParameter()`, but that doesn't give me the indirections?

Oh, yes. The API is a bit confusing here, I think. You can do `this.asParameter(1)` to mark the first indirection (i.e., `*argv`), whereas `this.asParameter()` (and `this.asParameter(0)`) marks `argv`.

**Comment by jketema** [comment]
Ah, yeah, `this.asParameter(_)` does the trick.

**Comment by jketema** [comment]
I propose we merge this one, and I open a separate PR that does the `this.asParameter(_)` replacement. So we can look at its effects in isolation?

**Comment by MathiasVP** [comment]
> Ah, yeah, this.asParameter(_) does the trick.

We should probably rename `this.asParameter(any number larger than zero)` to `this.asIndirectParameter(int n)`. (We should be careful to handle references properly like we do in the current `asParameter` implementation, though.)

**Comment by MathiasVP** [comment]
> I propose we merge this one, and I open a separate PR that does the this.asParameter(_) replacement. So we can look at its effects in isolation?

SGTM!

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Comment by jketema** [comment]
Opened https://github.com/github/codeql/pull/11693 as a draft PR.

**Pull request merged**